rfcs/45-stageless.md

```diff
@@ -567,44 +614,108 @@ fn main() {
-This means you can do `add_system(apply_system_buffers.after(X))` multiple times or use the `chain!` macro with multiple unnamed instances of the same function without having to name any of them.
-A system only needs a name when you want to do `before(name)` or `after(name)` somewhere else.
+This means you can do `add_system(apply_system_buffers.after(X))` multiple times or use the `chain!` macro with multiple unnamed instances of the same function without having to name any of them using their types. This protects against the error in (3) for many common cases.
+
+A system only needs to reference the potentially ambiguous `SystemTypeIdSet` when you want to do `before(name)` or `after(name)` somewhere else. And in these cases, the validation from (3) will protect users from accidentally doing something wrong.
+
+In short, when the error in (3) is encountered:
+
+```rust
+app
+    .add_system(foo)
+    .add_system(foo)
+    .add_system(bar.after(foo))
+```
+
+Users should do one of the following:
+
+1. Create a new set, add it to the duplicate system, and use that to define orders unambiguously;
+
+    ```rust
+    app
+        .add_system(foo)
+        .add_system(foo.in_set(X))
+        .add_system(bar.after(X))
+    ```
+
+2. Refactor their system registration to use APIs that order unambiguously using specific system instance ids:
+
+    ```rust
+    // chaining
+    app
+        .add_system(foo)
+        .add_system(chain![foo, bar])
+    // rotate bar.after(foo) to foo.before(bar)
+    app
+        .add_system(foo)
+        .add_system(bar)
+        .add_system(foo.before(bar))
+    ```
 
 *Note: In case it wasn't already clear, system chaining introduces new instances of systems. It does not reuse existing ones.*
 
 See the discussion on command-flushed ordering constraints and automatic inference of sync points in **Future Work** for ideas on how we can improve this.
 
+### If "system types" are System Sets, won't that allow users to configure them in confusing ways?
+
+Unless we do something to prevent this, then yes:
+
+```rust
+fn some_system(time: Res<Time>) {}
+
+app
+    .configure_set(some_system.after(X))
+    .add_system(foo.in_set(some_system))
+```
+
+This is certainly confusing! To resolve this, we've added the ability to "lock" sets:
+
+```rust
+app.configure_set(SomeSet.lock())
+```
+
+This will cause any further "set configuration" (adding ordering constraints, adding systems to sets, etc) from being added. SystemTypeIdSets (such as `some_system` in the example above) are automatically "locked" when they are initialized, preventing any user configuration.
+
+When adding systems, this lock flag will be internally ignored specifically for their automatically-computed `SystemTypeIdSet`. Ensuring that the set is still constructed correctly and ordering is respected.
+
+"Set locking" can also be used by Bevy plugin authors to make their sets "private" (much like "visibility" keywords in programming languages).

```

Comment:
This concept needs to be introduced in the main body.

Reply:
Totally onboard with this, but how are locks enforced if users can call `app.configure_set(P::X)` before `app.app_plugin(P)`?

I also think a very common situation will be: plugin wants to prevent users from adding nodes to their sets, but is OK with users adding conditions and dependencies. Having an enum to control how "strict" the lock is would be nice.

Reply:
Agreed. It feels like a concept of "exhaustive" is more important here than the broader concept of "locked".

Atomicity feels like the next level of guarantees you'll want after exhaustiveness.

Reply:
Yep.

If two plugins add sets to a schedule and their relative order is ambiguous (but that ambiguity is ignored), then their systems will likely be interleaved during execution. 

That random interleaving could break their logic even if the order between the sets didn't matter. But if we could say "this set should be treated as an atom", then the ambiguity would stay on the set-level.

Reply:
@maniwani made a good point about ordering. And as we discussed / learned in our discord discussion, locking is actually a larger, more controversial design problem than I initially thought. As we agreed in that discussion, I have removed locking from this RFC in favor of internal validation for SystemTypeIdSet, which is easier / less controversial to implement. We can expand into generalized locking later, if we need to.